crypto/thread/api.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/e_os2.h>
+#include <openssl/crypto.h>
+
+#if !defined(OPENSSL_THREADS)
+
+int CRYPTO_THREAD_enabled(OSSL_LIB_CTX *ctx)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_enable(OSSL_LIB_CTX *ctx, int max_threads)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_disable(OSSL_LIB_CTX *ctx)
+{
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OSSL_LIB_CTX *ctx, int max_threads)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_spawn_worker(OSSL_LIB_CTX *ctx, CRYPTO_WORKER_CALLBACK cb)
+{
+    return 0;
+}
+
+#else
+
+# include <internal/thread.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+static ossl_inline int thread_count_active(OSSL_LIB_CTX *ctx);
+
+int CRYPTO_THREAD_enabled(OSSL_LIB_CTX *ctx)
+{
+    int enabled;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
```

Comment:
Wouldn't an rwlock be more appropriate here?

Reply:
The task, worker and thread context locks are used with condvars and to control reads/writes, see https://github.com/ckalina/openssl/commit/59f08fb8fbcb322c840c03e25fde7254f88ee2a0.

I don't think I could use CRYPTO_RWLOCK here when USE_RWLOCK is defined. CRYPTO_CONDVAR_wait is a wrapper on top of pthread_cond_wait and SleepConditionVariableCS.

While I could use SleepConditionVariableSRW instead of SleepConditionVariableCS, I don't think there's a viable POSIX alternative for pthread_rwlock_t.

An alternative would be to have one mutex and one rwlock for each task, worker and thread context, but that alternative doesn't simplify the design too much.